### ISSUE TITLE
vpa from double seems overly slow

### ISSUE THREAD

**Issue by cbm755** [cbm755/octsympy#499]
````
>> x = rand(10,10);
>> 
>> 
>> tic; xv = vpa(x, 16); toc
Elapsed time is 15.0591 seconds.
>>  tic; sin(xv1); toc
Elapsed time is 0.109808 seconds.
>> 
````
Note it took way longer to *make* `xv` then to compute its sin...

**Comment by cbm755** [comment]
This is because of a dumb loop done in vpa.m.  We can fix it by passing all the doubles in at once and avoiding all those round-trips.

But it might be more fun to fix this with pytave where we can pass the input array to a numpy array all-at-once.

@genuinelucifer this is low priority but maybe interesting.

**Comment by genuinelucifer** [comment]
@cbm755 Sorry, I saw this now.
 I'll try it with and without pytave and do the same tests. I'll post the time taken in each case in a few hours. :)

**Comment by cbm755** [comment]
Cool.  I think it will still be slow, because it loops over the doubles to do that "magic_double_str" thing per element.  Could look at how `@sym/sym` deals with arrays of doubles, as IIRC its faster.

**Comment by genuinelucifer** [comment]
The difference between sym and vpa is only a factor of 2 for me.
x = [10 double values]
sym(x) took approx 0.4 seconds on average.
And vpa(x) took 0.8 seconds on average.
Also, sym(x) passes all the arguments in this case to another function numeric_array_to_sym which finally calls sym on each element.
While vpa calls itself on each element and then calls magic_double_str on that element.
I guess both approaches are nearly similar in terms of performance.
I tried changing a few other things but could only gain a performance boost of max 2.5% so I dropped those changes.
I will try a few other things. And see if we can bypass the magic_double_str thing somehow (atleast for most of the cases when we don't need it).

**Comment by cbm755** [comment]
That summary looks right to me, thanks for writing that down in one place.

I guess for vpa, people are more likely to want to take large double arrays and make them into vpa (for example, to repeat some algorithm with a smaller `eps`).  I think meaningful speed increase can only be gained by importing them all at once, either into a numpy array or at least into a python list..

Are you tests on pytave or master?  For me on master, a `rand(20,20)` array takes 33 s to convert to sym but 104 s to convert to vpa.  For `rand(1,10)` it was 0.8 s and 1.3s respectively.

**Comment by genuinelucifer** [comment]
I tried again (atleast 3 times each) using rand(20, 20) and following are the results (using '@' bookmark of mtmiller/pytave):  
.  
* On branch `pytave_ex`  :  
vpa took approx 140 seconds.  
sym took approx 20 seconds (a huge 7x difference!)  
BUT, I get the rat heuristics warning on sym. If I just comment out that warning line, there is significant improvement and the average time came down to around 13 seconds for sym!  
.  
* On branch `pytave`  :  
vpa took approx 160 seconds! (more than on pytave_ex, so our storing variables with pytave is worth something here!!)  
sym took approx 24 seconds  
If I comment out that warning line, time came down to around 15 seconds for sym.  
.
* On branch `master`  :  
vpa took approx 210 seconds! (much more than others)  
sym took approx 35 seconds! (much more than rest)  
If I comment out that warning line, time came down to around 25 seconds for sym.  
.  
So, to sum up, using more and more pytave helps us gain better performance. Also, that warning thing on sym creation is a real time consumer. We should probably only show this once when a list is passed as the argument.
